src/vm/types/serialization.rs

```diff
@@ -0,0 +1,862 @@
+use vm::errors::{RuntimeErrorType, InterpreterResult, InterpreterError};
+use vm::types::{Value, OptionalData, PrincipalData, TypeSignature, AtomTypeIdentifier, TupleData};
+use vm::database::{ClaritySerializable, ClarityDeserializable};
+
+use std::collections::HashMap;
+use serde_json::{Value as JSONValue};
+use util::hash;
+
+const TYPE_I128: &str = "i128";
+const TYPE_BOOL: &str = "bool";
+const TYPE_BUFF: &str = "buff";
+const TYPE_STANDARD_PRINCIPAL: &str = "principal";
+const TYPE_CONTRACT_PRINCIPAL: &str = "contract_principal";
+const TYPE_RESP_OK: &str = "ok";
+const TYPE_RESP_ERR: &str = "err";
+const TYPE_OPT_SOME: &str = "some";
+const TYPE_OPT_NONE: &str = "none";
+const TYPE_TUPLE: &str = "tuple";
+const TYPE_LIST: &str = "list";
+
+enum ContainerTypes { OPT_SOME, RESP_OK, RESP_ERR }
+
+#[derive(Deserialize)]
+#[serde(untagged, deny_unknown_fields)]
+enum JSONParser {
+    None {
+        #[serde(rename="type")]
+        type_n: String },
+    Bool { 
+        #[serde(rename="type")]
+        type_n: String,
+        value: bool },
+    Simple { 
+        #[serde(rename="type")]
+        type_n: String,
+        value: String },
+    Container {
+        #[serde(rename="type")]
+        type_n: String,
+        value: Box<JSONParser> },
+    List {
+        #[serde(rename="type")]
+        type_n: String,
+        entries: Vec<JSONParser> },
+    Tuple {
+        #[serde(rename="type")]
+        type_n: String,
+        entries: HashMap<String, JSONParser> },
+    ContractPrincipal {
+        #[serde(rename="type")]
+        type_n: String,
+        issuer: String,
+        name: String
+    }
+}
+
+fn to_hex(val: &i128) -> String {
+    if *val >= 0 {
+        format!("{:x}", val)
```

Comment:
Maybe this is a paranoid delusion on my part, but how stable can we expect formatting rules to be in practice?  I'm mainly concerned about whether or not the user's locale settings can affect the formatter's behavior (e.g. by making characters upper-case, by tacking on one or more leading 0's, by flipping the endianness, and so on).

Reply:
I share this fear, so it's at least a shared delusion.

Reply:
I spent a while poking at the LowerHex docs, but produced no satisfactory specification, so I'll just update this to do the hexing manually.

Reply:
Thank you for feeding my paranoia and general mistrust of GNU libc locale settings :ghost: